src/test/java/com/google/devtools/build/lib/server/GrpcServerTest.java

```diff
@@ -460,6 +469,320 @@ public void onCompleted() {
+            }
+          }
+        };
+    createServer(dispatcher);
+
+    CountDownLatch gotFoo = new CountDownLatch(1);
+    AtomicReference<RunResponse> lastFooResponse = new AtomicReference<>();
+    AtomicReference<RunResponse> lastBarResponse = new AtomicReference<>();
+
+    CommandServerStub stub = CommandServerGrpc.newStub(channel);
+    stub.run(
+        createPreemptibleRequest(firstCommandArg),
+        new StreamObserver<RunResponse>() {
+          @Override
+          public void onNext(RunResponse value) {
+            gotFoo.countDown();
+            lastFooResponse.set(value);
+          }
+
+          @Override
+          public void onError(Throwable t) {}
+
+          @Override
+          public void onCompleted() {}
+        });
+
+    // Wait for the first command to startup
+    gotFoo.await();
+
+    CountDownLatch gotBar = new CountDownLatch(1);
+    stub.run(
+        createPreemptibleRequest(secondCommandArg),
+        new StreamObserver<RunResponse>() {
+          @Override
+          public void onNext(RunResponse value) {
+            gotBar.countDown();
+            lastBarResponse.set(value);
+          }
+
+          @Override
+          public void onError(Throwable t) {}
+
+          @Override
+          public void onCompleted() {}
+        });
+
+    gotBar.await();
+    server.shutdown();
+    server.awaitTermination();
+
+    assertThat(lastBarResponse.get().getFinished()).isTrue();
+    assertThat(lastBarResponse.get().getExitCode()).isEqualTo(0);
+    assertThat(lastFooResponse.get().getFinished()).isTrue();
+    assertThat(lastFooResponse.get().getExitCode()).isEqualTo(8);
+    assertThat(lastFooResponse.get().hasFailureDetail()).isTrue();
+    assertThat(lastFooResponse.get().getFailureDetail().hasInterrupted()).isTrue();
+    assertThat(lastFooResponse.get().getFailureDetail().getInterrupted().getCode())
+        .isEqualTo(Code.INTERRUPTED);
+  }
+
+  /**
+   * Ensure that when a command is not marked as preemptible, running a second
+   * command does not interrupt the first command
+   */
+  @Test
+  public void testNoPreeempt() throws Exception {
+    String firstCommandArg = "Foo";
+    String secondCommandArg = "Bar";
+
+    AtomicInteger barsReceived = new AtomicInteger();
+
+    CommandDispatcher dispatcher =
+        new CommandDispatcher() {
+          @Override
+          public BlazeCommandResult exec(
+              InvocationPolicy invocationPolicy,
+              List<String> args,
+              OutErr outErr,
+              LockingMode lockingMode,
+              String clientDescription,
+              long firstContactTimeMillis,
+              Optional<List<Pair<String, String>>> startupOptionsTaggedWithBazelRc)
+              throws InterruptedException {
+            if (args.contains(firstCommandArg)) {
+              OutputStream out = outErr.getOutputStream();
+              try {
+                while (barsReceived.get() < 3) {
```

Comment:
Is the intention with the `3` here to generate some overlap? It took me a bit to figure this out, so I'd like to recommend instead gating it with latches to make it more explicit.

```
  CountDownLatch fooBlocked, fooProceed, barBlocked, BarProceed...

  ... exec(...) {
    if (foo) {
      fooBlocked.countDown(1);
      fooProceed.await();
    } else {
      barBlocked.countDown(1);
      barProceed.await();
    }
  }
  ...
  stub.run(fooRequest, ...);
  fooBlocked.await();

  stub.run(barRequest, ...);
  barBlocked.await();
  // At this point both commands should be blocked on proceed latch, carry on...
  fooProceed.countDown(1);
  barProceed.countDown(1);

  // do verifications that prove they weren't interrupted.
```

Reply:
Much better this way. Thanks